File: src/com/gmail/marcosav2010/communicator/packet/wrapper/PacketEncoder.java
package com.gmail.marcosav2010.communicator.packet.wrapper;

import java.io.Closeable;
import java.io.IOException;
import java.io.OutputStream;
import java.util.UUID;

import com.gmail.marcosav2010.communicator.packet.AbstractPacket;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class PacketEncoder implements Closeable, AutoCloseable {

	@Getter
	private final OutputStream out;
	private int length;

	public void write(boolean b) throws IOException {
		write((byte) (b ? 1 : 0));
	}

	public void write(byte b) throws IOException {
		addAndCheck(1);
		out.write(b);
	}

	public void write(short b) throws IOException {
		out.write(serialize(b, 2));
	}

	public void write(int b) throws IOException {
		out.write(serialize(b, 4));
	}

	public void write(long b) throws IOException {
		out.write(serialize(b, 8));
	}

	public void write(byte[] buf) throws IOException {
		addAndCheck(buf.length);
		write(buf.length);
		out.write(buf);
	}

	public void write(String s) throws IOException {
		write(s.getBytes("UTF-8"));
	}

	public void write(UUID uuid) throws IOException {
		write(uuid.getMostSignificantBits());
		write(uuid.getLeastSignificantBits());
	}

	private byte[] serialize(long o, int size) {
		addAndCheck(size);
		byte[] array = new byte[size];
		for (int i = 0; i < size; ++i)
			array[i] = (byte) (o >> 8 * (size - i - 1) & 0xFFL);
		return array;
	}

	private void addAndCheck(int i) {
		length += i;
		if (length > AbstractPacket.BASE_SIZE)
			throw new OverExceededByteLimitException(length);
	}

	@Override
	public void close() throws IOException {
		out.close();
	}
}

File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/wrapper/PacketEncoder.kt
package com.gmail.marcosav2010.communicator.packet.wrapper

import com.gmail.marcosav2010.communicator.packet.AbstractPacket
import com.gmail.marcosav2010.communicator.packet.wrapper.exception.OverExceededByteLimitException
import java.io.Closeable
import java.io.OutputStream
import java.nio.charset.StandardCharsets
import java.util.*

class PacketEncoder(private val out: OutputStream) : Closeable {

    private var length = 0

    fun write(b: Boolean) {
        write((if (b) 1 else 0).toByte())
    }

    fun write(b: Byte) {
        addAndCheck(1)
        out.write(b.toInt())
    }

    fun write(b: Short) {
        out.write(serialize(b.toLong(), 2))
    }

    fun write(b: Int) {
        out.write(serialize(b.toLong(), 4))
    }

    fun write(b: Long) {
        out.write(serialize(b, 8))
    }

    fun write(buf: ByteArray) {
        addAndCheck(buf.size)
        write(buf.size)
        out.write(buf)
    }

    fun write(s: String) {
        write(s.toByteArray(StandardCharsets.UTF_8))
    }

    fun write(uuid: UUID) {
        write(uuid.mostSignificantBits)
        write(uuid.leastSignificantBits)
    }

    private fun serialize(o: Long, size: Int): ByteArray {
        addAndCheck(size)
        val array = ByteArray(size)
        for (i in 0 until size) array[i] = (o shr 8 * (size - i - 1) and 0xFFL).toByte()
        return array
    }

    private fun addAndCheck(i: Int) {
        length += i
        if (length > AbstractPacket.BASE_SIZE) throw OverExceededByteLimitException(length.toLong())
    }

    override fun close() {
        out.close()
    }
}